Feature/joke widget

brody-ly opened the pull request:
This pull request adds a Joke widget suggested in #483. Its powered by [jokeapi](https://jokeapi.dev/).

The Joke widget options include:

- A selection of supported categories from jokeapi
- Refreshing the joke every x minutes,hours,days
- Displaying single and two part jokes 

Screenshot of options:
![image](https://user-images.githubusercontent.com/26295699/166186202-2b33cde5-94c4-4a82-a71a-bbced4fee3d4.png)

Jokes are categorised into two types, single joke and two-part jokes. Single jokes will display as is whereas two-part jokes will display differently depending on whether it is a question or not. 

If the two-part joke is a question you'll need to click the question to reveal the answer otherwise the delivery and setup will be separated with . . .

Example of a single joke:
![image](https://user-images.githubusercontent.com/26295699/166185910-b4c7cd28-293f-4c22-bf2f-61da36c440d9.png)

Example of two-part joke:
![image](https://user-images.githubusercontent.com/26295699/166186050-1f1bd29f-9564-4940-a869-38ae0b70f6d7.png)

Example of a two-part joke that's a question: 
![image](https://user-images.githubusercontent.com/26295699/166185797-ca66c400-9af2-4a8e-bee0-0a73e6c4c99a.png)

Would love to get some feedback on this PR! 

Thanks

Yohn commented:
This is awesome!! Thanks for diving it and building a PR for this!

My only feedback would be if there is a way to set the blacklist flags option that's available? Some of the jokes might "trigger" some of the users, and I'd hate to have it removed because of that.
![Tabliss-Jokes-FlagOptions](https://user-images.githubusercontent.com/2002591/166516486-f91d3d35-6e9e-4bc7-b063-8a83ec3b4466.png)

joelshepherd commented on src/plugins/widgets/joke/Joke.tsx:
```diff
@@ -0,0 +1,87 @@
+    return (
+      <>
+        <h5>{joke.setup}</h5>
+        <p>. . .</p>
```
Does it look good without this?

joelshepherd commented on src/plugins/widgets/joke/JokeSettings.tsx:
```diff
@@ -0,0 +1,103 @@
+            setData({ ...data, timeout: Number(event.target.value) })
+          }
+        >
+          <option value={5 * MINUTES * 1000}>Every 5 minutes</option>
+          <option value={15 * MINUTES * 1000}>Every 15 minutes</option>
```
```suggestion
          <option value={5 * MINUTES}>Every 5 minutes</option>
          <option value={15 * MINUTES}>Every 15 minutes</option>
```

joelshepherd reviewed (commented):
Thanks for the contribution! Overall looks great.

We probably need to remove the "NSFW" toggle and force safe-mode. Jokes that are too dark would probably void Tabliss from [extension stores](https://developer.chrome.com/docs/webstore/program_policies/#hate-speech) :(

Last suggestion, you could try and match the user's language preference against languages supported by the jokes API.

```ts
// Grab the user's locale
const locale = useValue(db, "locale");

// Map to supported language
const mapLocaleToAPILang = (locale: string): string => {
  const supported = ["en", "de", ...];
  const [lang] = locale.split("-");
  return supported.includes(lang) ? lang : "en";
}
```

brody-ly commented:
>We probably need to remove the "NSFW" toggle and force safe-mode. Jokes that are too dark would probably void Tabliss from [extension stores](https://developer.chrome.com/docs/webstore/program_policies/#hate-speech) :(

Good call on this, it completely slipped my mind. I've removed NSFW and Dark toggle and have updated the API call to only return safe jokes. 

One thing to note about [JokeApi's safe mode](https://jokeapi.dev/#safe-mode) is that it's only a best-effort filter, meaning there could be an unsafe joke that might slip through. Might be worth re-considering if this feature is worth putting in, but totally your call @joelshepherd. 

> Last suggestion, you could try and match the user's language preference against languages supported by the jokes API.

Thanks for the suggestion! I've added language support now.

Thanks for the review and the awesome tab extension!

joelshepherd commented:
Looks great to me.

> Might be worth re-considering if this feature is worth putting in

I think we've made appropriate steps to avoid violating the policies with the safe mode, so happy to roll with it.

Thanks for the addition!
